File: Data Structures & Algorithms/count-connected-components/submission-2.py
class Solution:
    def countComponents(self, n: int, edges: List[List[int]]) -> int:
        graph = [[] for _ in range(n)]

        for edge in edges:
            graph[edge[0]].append(edge[1])
            graph[edge[1]].append(edge[0])
        
        visited = [False] * n

        def dfs(node: int):
            visited[node] = True

            for neighbor in graph[node]:
                if visited[neighbor]:
                    continue
                
                dfs(neighbor)
        
        count = 0

        for node in range(n):
            if visited[node]:
                continue
            count += 1
            dfs(node)
        
        return count

File: Data Structures & Algorithms/count-paths/submission-2.go
func uniquePaths(m int, n int) int {
	memo := make([][]int, m)
	for i := range memo {
		memo[i] = make([]int, n)
	}

	var dfs func(i, j int) int

	dfs = func(i, j int) int {
		if i >= m || j >= n {
			return 0
		}

		if i == m-1 && j == n-1 {
			return 1
		}

		if memo[i][j] != 0 {
			return memo[i][j]
		}

		memo[i][j] = dfs(i+1, j) + dfs(i, j+1)

		return memo[i][j]
	}

	return dfs(0, 0)
}
